File: leetcode/medium/1382. Balance a Binary Search Tree/solution.js
// https://leetcode.com/problems/balance-a-binary-search-tree/
class TreeNode {
    constructor(val, left, right) {
        this.val = val ?? 0;
        this.left = left ?? null;
        this.right = right ?? null;
    }
}

/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
function balanceBST(root) {
    const items = [];

    const inorderTraverse = (node) => {
        if (!node.val) return;
        items.push(node.val);
        if (node.left) inorderTraverse(node.left);
        if (node.right) inorderTraverse(node.right);
    };

    inorderTraverse(root);
    items.sort((a, b) => a - b);
    const makeBalanced = (begin, end) => {
        if (begin > end) return null;
        const mid = Math.floor((begin + end) / 2);
        const node = new TreeNode(items[mid]);
        node.left = makeBalanced(begin, mid - 1);
        node.right = makeBalanced(mid + 1, end);
        return node;
    };

    const resNode = makeBalanced(0, items.length - 1);
    return resNode;
};

const treenode = new TreeNode(1);
treenode.left = null;
treenode.right = new TreeNode(2);

balanceBST(treenode);

File: leetcode/medium/1382. Balance a Binary Search Tree/solution.cc
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<TreeNode*> sorted;
    TreeNode* balanceBST(TreeNode* root) {
        inorderTraverse(root);
        return vector2BST(0, sorted.size() - 1);
    }

    void inorderTraverse(TreeNode* root) {
        if (root == NULL) return;
        inorderTraverse(root->left);
        sorted.push_back(root);
        inorderTraverse(root->right);
    }

    TreeNode* vector2BST(int begin, int end) {
        if (begin > end) return NULL;
        int mid = (begin + end) / 2;
        TreeNode* root = sorted[mid];
        root->left = vector2BST(begin, mid - 1);
        root->right = vector2BST(mid + 1, end);
        return root;
    }
};

int main() {
    return 0;
}